include/simulator/bug_notifier.hpp

```diff
@@ -17,7 +17,7 @@ class Bug;  // Forward declaration.
  */
 class BugNotifier {
  private:
-  std::set<std::shared_ptr<BugObserver>> observers_;
+  std::unordered_set<BugObserver*> observers_;
```

Comment:
Because of the issue talked offline about creating a `shared_ptr` to `this`. 
I know we can kind of trick this by using the tools we discussed offline as well, however, I don't thing it's a good practice to use so here. Usually, the creation of the content held by a `shared_ptr` represents ownership, which is not the case here (as the class does not own itself). By doing some research I found out that, even in cpp, the usual way to code this pattern is using raw pointers. I don't want to get to philosophical here, but the issue with raw pointers are related to memory allocation, and here we're not creating and/or destructing anything, so it's the perfect use case for raw pointers to me.